File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/sms/forward/SMSForwardViewModel.kt
package com.openu60.feature.sms.forward

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.*
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SMSForwardState(
    val config: SmsForwardConfig = SmsForwardConfig(),
    val log: List<ForwardLogEntry> = emptyList(),
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class SMSForwardViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(SMSForwardState())
    val state: StateFlow<SMSForwardState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSON("/api/sms/forward/config")
                val config = SMSForwardParser.parseConfig(data)
                _state.value = _state.value.copy(config = config, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun updateConfig(
        enabled: Boolean,
        pollIntervalSecs: Long,
        markRead: Boolean,
        deleteAfter: Boolean,
    ) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/sms/forward/config", mapOf(
                    "enabled" to enabled,
                    "poll_interval_secs" to pollIntervalSecs,
                    "mark_read_after_forward" to markRead,
                    "delete_after_forward" to deleteAfter,
                ))
                _state.value = _state.value.copy(
                    config = _state.value.config.copy(
                        enabled = enabled,
                        pollIntervalSecs = pollIntervalSecs,
                        markReadAfterForward = markRead,
                        deleteAfterForward = deleteAfter,
                    ),
                    isLoading = false,
                    message = "Config saved",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) updateConfig(enabled, pollIntervalSecs, markRead, deleteAfter)
                else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun toggleEnabled(enabled: Boolean) {
        val previous = _state.value.config.enabled
        _state.value = _state.value.copy(
            config = _state.value.config.copy(enabled = enabled)
        )
        viewModelScope.launch {
            try {
                agentClient.putJSON("/api/sms/forward/config", mapOf(
                    "enabled" to enabled,
                    "poll_interval_secs" to _state.value.config.pollIntervalSecs,
                    "mark_read_after_forward" to _state.value.config.markReadAfterForward,
                    "delete_after_forward" to _state.value.config.deleteAfterForward,
                ))
                // Re-assert in case refresh() overwrote during coroutine suspension
                _state.value = _state.value.copy(
                    config = _state.value.config.copy(enabled = enabled)
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) toggleEnabled(enabled)
                else revertEnabled(previous, e.message)
            } catch (e: Exception) {
                revertEnabled(previous, e.message)
            }
        }
    }

    private fun revertEnabled(previous: Boolean, msg: String?) {
        _state.value = _state.value.copy(
            config = _state.value.config.copy(enabled = previous),
            message = msg ?: "Failed to toggle",
            messageIsError = true,
        )
    }

    fun createRule(name: String, filter: SmsFilter, destination: ForwardDestination) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/sms/forward/rules", mapOf(
                    "name" to name,
                    "filter" to SMSForwardParser.filterToMap(filter),
                    "destination" to SMSForwardParser.destinationToMap(destination),
                ))
                _state.value = _state.value.copy(
                    message = "Rule created",
                    messageIsError = false,
                )
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) createRule(name, filter, destination)
                else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun updateRule(id: Int, name: String, enabled: Boolean, filter: SmsFilter, destination: ForwardDestination) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.putJSON("/api/sms/forward/rules", mapOf(
                    "id" to id,
                    "name" to name,
                    "enabled" to enabled,
                    "filter" to SMSForwardParser.filterToMap(filter),
                    "destination" to SMSForwardParser.destinationToMap(destination),
                ))
                _state.value = _state.value.copy(
                    message = "Rule updated",
                    messageIsError = false,
                )
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) updateRule(id, name, enabled, filter, destination)
                else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun deleteRule(id: Int) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.deleteJSON("/api/sms/forward/rules", mapOf("id" to id))
                _state.value = _state.value.copy(message = "Rule deleted", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) deleteRule(id) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun toggleRule(id: Int, enabled: Boolean) {
        viewModelScope.launch {
            _state.value = _state.value.copy(message = null)
            try {
                agentClient.putJSON("/api/sms/forward/rules/toggle", mapOf(
                    "id" to id,
                    "enabled" to enabled,
                ))
                _state.value = _state.value.copy(
                    config = _state.value.config.copy(
                        rules = _state.value.config.rules.map {
                            if (it.id == id) it.copy(enabled = enabled) else it
                        },
                    ),
                    message = "Rule ${if (enabled) "enabled" else "disabled"}",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) toggleRule(id, enabled) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun testDestination(destination: ForwardDestination) {
        viewModelScope.launch {
            _state.value = _state.value.copy(message = null)
            try {
                agentClient.postJSON("/api/sms/forward/test", mapOf(
                    "destination" to SMSForwardParser.destinationToMap(destination),
                ))
                _state.value = _state.value.copy(
                    message = "Test message sent successfully",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) testDestination(destination) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun fetchLog() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSONArray("/api/sms/forward/log")
                val log = data.mapNotNull { SMSForwardParser.parseLogEntry(it) }
                _state.value = _state.value.copy(log = log, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) fetchLog() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun clearLog() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/sms/forward/log/clear")
                _state.value = _state.value.copy(
                    log = emptyList(),
                    isLoading = false,
                    message = "Log cleared",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) clearLog() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}

File: mobile/ios/OpenU60/Features/SMS/Forward/SMSForwardViewModel.swift
import SwiftUI

@Observable
@MainActor
final class SMSForwardViewModel {
    var config = SmsForwardConfig()
    var lastForwardedId: Int = 0
    var log: [ForwardLogEntry] = []
    var isLoading = false
    var message: String?
    var messageIsError = false
    var presentedSheet: Sheet?

    enum Sheet: Identifiable {
        case add
        case edit(ForwardRule)
        var id: String {
            switch self {
            case .add: "add"
            case .edit(let r): "edit-\(r.id)"
            }
        }
    }

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    // MARK: - Config

    func refresh() async {
        isLoading = true
        message = nil
        do {
            let data = try await client.getJSON("/api/sms/forward/config")
            if let configDict = data["config"] as? [String: Any] {
                config = SMSForwardParser.parseConfig(configDict)
            }
            lastForwardedId = (data["last_forwarded_id"] as? Int)
                ?? (data["last_forwarded_id"] as? NSNumber)?.intValue
                ?? 0
        } catch {
            showMessage("Failed to load: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func updateConfig(enabled: Bool, pollIntervalSecs: Int, markRead: Bool, deleteAfter: Bool) async {
        isLoading = true
        do {
            let body: [String: Any] = [
                "enabled": enabled,
                "poll_interval_secs": pollIntervalSecs,
                "mark_read_after_forward": markRead,
                "delete_after_forward": deleteAfter,
            ]
            let _ = try await client.putJSON("/api/sms/forward/config", body: body)
            config.enabled = enabled
            config.pollIntervalSecs = pollIntervalSecs
            config.markReadAfterForward = markRead
            config.deleteAfterForward = deleteAfter
            showMessage("Settings saved", isError: false)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func toggleEnabled(_ enabled: Bool) async {
        let previous = config.enabled
        config.enabled = enabled  // optimistic
        do {
            let _ = try await client.putJSON("/api/sms/forward/config", body: [
                "enabled": enabled,
                "poll_interval_secs": config.pollIntervalSecs,
                "mark_read_after_forward": config.markReadAfterForward,
                "delete_after_forward": config.deleteAfterForward,
            ])
            config.enabled = enabled  // re-assert in case refresh() overwrote during await
        } catch {
            config.enabled = previous  // revert on failure
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
    }

    // MARK: - Rules

    func createRule(name: String, filter: SmsFilter, destination: ForwardDestination) async {
        isLoading = true
        do {
            var body: [String: Any] = [
                "name": name,
                "filter": SMSForwardParser.filterToDict(filter),
                "destination": SMSForwardParser.destinationToDict(destination),
            ]
            body["enabled"] = true
            let _ = try await client.postJSON("/api/sms/forward/rules", body: body)
            showMessage("Rule created", isError: false)
            await refresh()
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func updateRule(id: Int, name: String, enabled: Bool, filter: SmsFilter, destination: ForwardDestination) async {
        isLoading = true
        do {
            let body: [String: Any] = [
                "id": id,
                "name": name,
                "enabled": enabled,
                "filter": SMSForwardParser.filterToDict(filter),
                "destination": SMSForwardParser.destinationToDict(destination),
            ]
            let _ = try await client.putJSON("/api/sms/forward/rules", body: body)
            showMessage("Rule updated", isError: false)
            await refresh()
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func deleteRule(id: Int) async {
        do {
            let _ = try await client.deleteJSON("/api/sms/forward/rules", body: ["id": id])
            config.rules.removeAll { $0.id == id }
            showMessage("Rule deleted", isError: false)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
    }

    func toggleRule(id: Int, enabled: Bool) async {
        do {
            let _ = try await client.putJSON("/api/sms/forward/rules/toggle", body: ["id": id, "enabled": enabled])
            if let idx = config.rules.firstIndex(where: { $0.id == id }) {
                config.rules[idx].enabled = enabled
            }
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
    }

    // MARK: - Test

    func testDestination(_ destination: ForwardDestination) async {
        isLoading = true
        do {
            let body: [String: Any] = [
                "destination": SMSForwardParser.destinationToDict(destination),
            ]
            let _ = try await client.postJSON("/api/sms/forward/test", body: body)
            showMessage("Test message sent", isError: false)
        } catch {
            showMessage("Test failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    // MARK: - Log

    func fetchLog() async {
        isLoading = true
        do {
            let data = try await client.getJSONArray("/api/sms/forward/log")
            log = data.compactMap { SMSForwardParser.parseLogEntry($0) }
        } catch {
            showMessage("Failed to load log: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func clearLog() async {
        do {
            let _ = try await client.postJSON("/api/sms/forward/log/clear")
            log = []
            showMessage("Log cleared", isError: false)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
    }

    func retryForward(index: Int) async {
        do {
            let _ = try await client.postJSON("/api/sms/forward/retry", body: ["index": index])
            showMessage("Retry succeeded", isError: false)
            await fetchLog()
        } catch {
            showMessage("Retry failed: \(error.localizedDescription)", isError: true)
            await fetchLog()
        }
    }

    // MARK: - Helpers

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}
